lib/executor/ramping_vus.go

```diff
@@ -458,8 +457,8 @@ func (vlvc RampingVUsConfig) GetExecutionRequirements(et *lib.ExecutionTuple) []
 		steps = vlvc.reserveVUsForGracefulRampDowns(steps, executorEndOffset)
 	}
 
-	// If the last PlannedVUs value wasn't 0, add a last step with 0
-	if steps[len(steps)-1].PlannedVUs != 0 {
+	// add one step for the end of the gracefulStop
+	if steps[len(steps)-1].TimeOffset != executorEndOffset {
```

Comment:
Hmm I am not sure this check is correct now :thinking: Shouldn't we always add this last step with 0 VUs here? That is, I think we should remove the `if` and just `append()` the step always :thinking:

Reply:
To see how this fails, try running this valid script:
```js
import http from 'k6/http';
import exec from 'k6/execution';
import { sleep } from 'k6';

export const options = {
    scenarios: {
        test: {
            executor: 'ramping-vus',
            startVUs: 0,
            stages: [
                { duration: '2s', target: 2 },
            ],
            gracefulRampDown: '0s',
            gracefulStop: '0s',
        },
    },
};

export default function () {
    console.log(`[t=${(new Date()) - exec.scenario.startTime}ms] VU#${exec.vu.idInTest} ran scenario iteration ${exec.scenario.iterationInTest} (${exec.vu.iterationInScenario} for VU)`);
    sleep(1);
}
```

You'll get this with your branch:
```
ERRO[0000] Executor error                                error="test expected graceful end offset at 2s to be final" executor=test startTime=0s type=ramping-vus
ERRO[0000] test expected graceful end offset at 2s to be final 
exit status 103
```

that is, this check gets triggered: https://github.com/grafana/k6/blob/27f2ead359dac24b4131dac4b57c95c058f06e48/lib/executor/ramping_vus.go#L516

Reply:
Mind you, while I am pretty sure this code is wrong, I am far from sure my proposal is the correct one... :sweat_smile:

Reply:
I am pretty sure that removing this `if` broke something else :rofl: I just don't remember what currently. But I did just try to add more steps with the motto of "it can't be worse" and it turned out it was breaking stuff ... similar to the breakage you are showing :(

Reply:
Quite possible... :disappointed: I'll try to debug this some more tomorrow, it's taking me a bit of time to grok what exactly was the logic behind everything here... :sweat_smile:

Reply:
@mstoykov, can you try to remember what gets broken when the `if` got removed, since I just tried it and the tests seem to pass fine :confused: 

Alternatively, I think we can use the following `if` instead:
```suggestion
	if steps[len(steps)-1].PlannedVUs != 0 || steps[len(steps)-1].TimeOffset != executorEndOffset {
```

The tests also pass with it too and the logic is that if one of the two conditions for the final step is not satisfied, we add a new final step after it the expected time offset and 0 value.

Reply:
So yeah, my change suggestion would be to fix the `if` in the way :arrow_up: and to add a `ramping-vus` scenario with this config to the tests:
```js
{
    // ...
    stages: [
        { duration: '2s', target: 2 },
    ],
    gracefulRampDown: '0s',
    gracefulStop: '0s',
}
```

Reply:
Can't find anything wrong with removing it as a whole as well. Maybe I tested with some other change  as well or am misremembering

Reply:
:+1: 

Still, I think I slightly prefer having the `if` there, just with both the `PlannedVUs` and `TimeOffset` conditions. Without the `if`, the last step could be repeated twice, which _shouldn't_ be a problem, but it also shouldn't be necessary :man_shrugging:

Reply:
I will add a test case for the script you posted above and expand the `if`